portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -154,4 +170,35 @@ private int getNumQuoteDisplayed(HttpServletRequest request, int defaultNumToDis
       return defaultNumToDisplay; 
     }
   }
+
+  /**
+   * Extracts the attributes of a quoteEntity and puts them into a new
+   * instance of Quote. 
+   */ 
+  private Quote extractQuoteFromEntity(Entity quoteEntity) {
+    long quoteId = quoteEntity.getKey().getId();
+    String text = (String) quoteEntity.getProperty(TEXT);
+    long timestampMillis = (long) quoteEntity.getProperty(TIMESTAMP);
+    String userEmail = (String) quoteEntity.getProperty(USER_EMAIL); 
+    String nickname = (String) quoteEntity.getProperty(NICKNAME); 
+    Quote newQuote = new Quote(quoteId, text, timestampMillis, userEmail, nickname);
+    return newQuote; 
+  } 
+
+  /**
+   * Returns the nickname of the user with id, or null if the user has not set a nickname.
+   */
+  private String getUserNickname(String userId) {
+    DatastoreService datastore = getDatastoreServiceWithConsistency();
+    Query query =
+        new Query("UserInfo")
+            .setFilter(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, userId));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return null;
```

Comment:
This is a good use case to use `java.util.Optional` instead of returning `null`. `Optional` lets you represent a value that might not be there, similar to how `null` works but it's more explicit, so it signals to the caller (using the type system) that they should think about how to handle the empty case, whereas right now it's not obvious to the caller that this method might return null (even if it's in the javadoc).

`Optional` has methods like `.ifPresent(...)` that let you do something with the value inside, or you can explicitly check with `.isPresent()` and then call `.get()`.

You can create an `Optional` with `Optional.ofNullable(...)`.

Reply:
I incorporated Optional in this method. In other methods in other classes where I need to fetch a nickname, I also use an Optional and check if it holds a value.